File: api/src/main/java/com/leandroadal/vortasks/entities/backup/userprogress/Achievement.java
package com.leandroadal.vortasks.entities.backup.userprogress;

import com.leandroadal.vortasks.entities.backup.Backup;
import com.leandroadal.vortasks.entities.backup.userprogress.dto.AchievementDTO;
import jakarta.persistence.CascadeType;
import jakarta.persistence.Entity;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.JoinColumn;
import jakarta.persistence.ManyToOne;
import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

@Getter
@Setter
@Entity
@NoArgsConstructor
@AllArgsConstructor
public class Achievement {

    @Id
    @GeneratedValue(strategy = GenerationType.UUID)
    private String id; 

    private String title;
    private String description;
    private Integer coins;
    private Integer gems;
    private boolean unlocked;

    @ManyToOne(cascade = CascadeType.ALL)
    @JoinColumn(name = "backup_id")
    private Backup userBackup;


    public void edit(AchievementDTO achievementDTO) {
        this.title = achievementDTO.title();
        this.description = achievementDTO.description();
        this.coins = achievementDTO.coins();
        this.gems = achievementDTO.gems();
        this.unlocked = achievementDTO.unlocked();
    }
}

File: app/vortasks/lib/models/achievement/achievement.dart
class Achievement {
  final String id;
  final String title;
  final String description;
  final int coins;
  final int gems;
  bool unlocked;

  Achievement({
    required this.id,
    required this.title,
    required this.description,
    required this.coins,
    required this.gems,
    this.unlocked = false,
  });

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'title': title,
      'description': description,
      'coins': coins,
      'gems': gems,
      'unlocked': unlocked,
    };
  }

  factory Achievement.fromJson(Map<String, dynamic> json) {
    return Achievement(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      coins: json['coins'],
      gems: json['gems'],
      unlocked: json['unlocked'],
    );
  }
}
